nexus/db-queries/src/db/pool_connection.rs

```diff
@@ -4,46 +4,128 @@
 
 //! Customization that happens on each connection as they're acquired.
 
+use anyhow::anyhow;
+use async_bb8_diesel::AsyncR2D2Connection;
 use async_bb8_diesel::AsyncSimpleConnection;
-use async_bb8_diesel::Connection;
-use async_bb8_diesel::ConnectionError;
 use async_trait::async_trait;
-use bb8::CustomizeConnection;
+use diesel::Connection;
 use diesel::PgConnection;
 use diesel_dtrace::DTraceConnection;
+use qorb::backend::{self, Backend, Error};
+use slog::Logger;
 
 pub type DbConnection = DTraceConnection<PgConnection>;
 
 pub const DISALLOW_FULL_TABLE_SCAN_SQL: &str =
     "set disallow_full_table_scans = on; set large_full_scan_rows = 0;";
 
-/// A customizer for all new connections made to CockroachDB, from Diesel.
-#[derive(Debug)]
-pub(crate) struct ConnectionCustomizer {}
+/// A [backend::Connector] which provides access to [PgConnection].
+pub(crate) struct DieselPgConnector {
+    log: Logger,
+    prefix: String,
+    suffix: String,
+}
+
+pub(crate) struct DieselPgConnectorArgs<'a> {
+    pub(crate) user: &'a str,
+    pub(crate) db: &'a str,
+    pub(crate) args: Option<&'a str>,
+}
 
-impl ConnectionCustomizer {
-    pub(crate) fn new() -> Self {
-        Self {}
+impl DieselPgConnector {
+    /// Creates a new "connector" to a database, which
+    /// swaps out the IP address at runtime depending on the selected backend.
+    ///
+    /// Format of the url is:
+    ///
+    /// - postgresql://{user}@{address}/{db}
+    ///
+    /// Or, if arguments are supplied:
+    ///
+    /// - postgresql://{user}@{address}/{db}?{args}
+    pub(crate) fn new(log: &Logger, args: DieselPgConnectorArgs<'_>) -> Self {
+        let DieselPgConnectorArgs { user, db, args } = args;
+        Self {
+            log: log.clone(),
+            prefix: format!("postgresql://{user}@"),
+            suffix: format!(
+                "/{db}{}",
+                args.map(|args| format!("?{args}")).unwrap_or("".to_string())
+            ),
+        }
     }
 
-    async fn disallow_full_table_scans(
-        &self,
-        conn: &mut Connection<DbConnection>,
-    ) -> Result<(), ConnectionError> {
-        conn.batch_execute_async(DISALLOW_FULL_TABLE_SCAN_SQL).await?;
-        Ok(())
+    fn to_url(&self, address: std::net::SocketAddr) -> String {
+        format!(
+            "{prefix}{address}{suffix}",
+            prefix = self.prefix,
+            suffix = self.suffix,
+        )
```

Comment:
- https://crates.io/crates/uri looks deprecated, so probably not that?
- https://docs.rs/url/2.5.2/url/ looks popular , and like it could be used for this purpose, though it doesn't have exactly that API you described
- https://docs.rs/http/latest/http/uri/struct.Builder.html also looks like it's a builder for URLs

I'm gonna go with the `url` crate, and use it primarily for the query parameters.